staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -162,144 +162,173 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+	}
+
+	if schema.Ref != nil {
+		openapiRef, err := spec.NewRef(*schema.Ref)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ref"), *schema.Ref, err.Error()))
+		}
+
+		if !openapiRef.IsValidURI() {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ref"), *schema.Ref, "ref does not point to a valid URI."))
+		}
+	}
+
+	allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISpec(schema.Not, fldPath.Child("not"), specVersion)...)
+
+	if len(schema.AllOf) != 0 {
+		for _, jsonSchema := range schema.AllOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISpec(&jsonSchema, fldPath.Child("allOf"), specVersion)...)
+		}
+	}
+
+	if len(schema.OneOf) != 0 {
+		for _, jsonSchema := range schema.OneOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISpec(&jsonSchema, fldPath.Child("oneOf"), specVersion)...)
+		}
+	}
+
+	if len(schema.AnyOf) != 0 {
+		for _, jsonSchema := range schema.AnyOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISpec(&jsonSchema, fldPath.Child("anyOf"), specVersion)...)
+		}
+	}
+
+	if len(schema.Properties) != 0 {
+		for property, jsonSchema := range schema.Properties {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISpec(&jsonSchema, fldPath.Child("properties."+property), specVersion)...)
+		}
+	}
+
+	if schema.Items != nil {
+		allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISpec(schema.Items.Schema, fldPath.Child("items"), specVersion)...)
+		if len(schema.Items.JSONSchemas) != 0 {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("items"), "items must be a schema object and not an array"))
+		}
+	}
+
+	return allErrs
+}
+
+type additionalSpecValidatorV2 struct{}
+
+// validateCustomResourceDefinitionOpenAPISpec validates against OpenAPI Spec v2.
+func (v *additionalSpecValidatorV2) validateCustomResourceDefinitionOpenAPISpec(schema *apiextensions.JSONSchemaProps, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if schema == nil {
+		return allErrs
+	}
+
+	if len(schema.OneOf) != 0 {
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("oneOf"), "oneOf is not supported in OpenAPI Spec v2."))
 	}
 
-	if len(OpenAPISpecV2.AnyOf) != 0 {
+	if len(schema.AnyOf) != 0 {
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("anyOf"), "anyOf is not supported in OpenAPI Spec v2."))
 	}
 
-	if OpenAPISpecV2.Not != nil {
+	if schema.Not != nil {
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("not"), "not is not supported in OpenAPI Spec v2."))
 	}
 
-	allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISpec(OpenAPISpecV2, fldPath.Child(""))...)
+	openAPISpecV3 := &additionalSpecValidatorV3{}
+	allErrs = append(allErrs, openAPISpecV3.validateCustomResourceDefinitionOpenAPISpec(schema, fldPath)...)
```

Comment:
If the checks applied to both v2 and v3, why don't we do them in the common validator function above? In theory, the additional validator for v3 should be empty.